sdk/program/PROJECT.md

```diff
@@ -0,0 +1,138 @@
+some priorities and a plan to make things better.
+
+## Concepts for the Program SDK
+
+Here are the priorities for the future of the SDK, in order of importance.
+
+1. Secure
+2. Respect semantic versioning
+3. On-chain program developers come first
+4. Possible to build a new program quickly and easily
+5. Sensible defaults, but allow customization
+6. Minimal dependencies
+7. Kitchen sinks only re-export functionality
+8. Default to new crates or files to bundle functionality
+9. If new functionality can't be done in a new crate or file, use features
+10. Easy to publish
+11. Easy for Agave devs to make changes
+
+## Project Plan Phase 1: solana_program updates
+
+### Break it up: if it can reasonably be used independently, it gets its own crate
+
+As one approach to address point number 4 of easily allowing customization, we
+should avoid bundling too much functionality in solana-program, and instead
+break it up sensibly. This way, someone can easily drop in a custom entrypoint
+without pulling in the default during build time.
+
+This also helps address point number 8 of avoiding kitchen sinks. By putting
+only related functionality into separate crates, there is no temptation for
+developers to "just throw it into the SDK." Instead, they'll have to consider
+the best place for their changes.
+
+When breaking up solana-program, we will take it in phases. First, we'll extract
+functionality required for program clients, which typically involve creating
+instructions and deserializing types. Next, we will actually create program
+client crates for the native programs. And finally, we will extract functionality
+used for writing on-chain programs.
+
+Here's a more specific breakdown:
+
+* Add crates required for program clients:
+  * borsh
+  * clock
+  * example-mocks
+  * hash
+  * instruction
+  * message
+  * program-error
+  * pubkey
+  * serde-varint
+  * sysvar
+* Those crates allow us to make program client crates:
+  * address lookup table
+  * bpf loader upgradeable
+  * feature
+  * loader v4
+  * stake
+  * system
+  * vote
+* Next, focus on crates for writing programs:
+  * account-info
+  * entrypoint
+  * program-pack
+  * program-syscalls
+  * stable-layout
+* Finally, do all the rest:
+  * native-token
+  * serialize-utils
+  * wasm
+* Deprecate any bits that make sense
+* `solana_program` becomes a crate with only re-exports and deprecated code
+
+### Add crate features
+
+After solana-program has been broken up, the next bit is to address point number
+7 and minimize dependencies. This will be done 
+
+* Serialization: serde / borsh / bincode / bytemuck optional
+* Syscalls: `#[cfg(not(target_os = "solana"))]` implementations of syscalls (ie.
+`hashv`) pull in unneeded dependencies. Where possible, either gate these deps
+with `[target.'cfg(not(target_os = "solana"))'.dependencies]` or create features
+like `backup-syscall`
+* Add an `unstable` feature for new code that could break, similar to Rust
+
+### Update CI
```

Comment:
I'd add [cargo hack](https://github.com/taiki-e/cargo-hack) here: it can test the power set of features provided each crate has a small-ish number of features (since power set is 2^N)